### PR TITLE
allow reads with zero refbases

### PR DESCRIPTION
**Pull request by akiezun**
addresses problem detected when working on #1413

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java [hunk 1]
```diff
@@ -40,7 +40,8 @@
 
     private BucketUtils(){} //private so that no one will instantiate this class
 
-    public static boolean isCloudStorageUrl(String path) {
+    public static boolean isCloudStorageUrl(final String path) {
+        Utils.nonNull(path);
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java]: unrelated to the core of this PR but a useful check

File: src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java [hunk 2]
```diff
@@ -101,6 +102,7 @@ public static InputStream openFile(String path, PipelineOptions popts) {
      * @return an OutputStream that writes to the specified file.
      */
     public static OutputStream createFile(String path, PipelineOptions popts) {
+        Utils.nonNull(path);
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java]: unrelated to the core of this PR but a useful check

File: src/test/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibraryUnitTest.java [hunk 3]
```diff
@@ -211,7 +211,7 @@ public void failsValidAlignmentEndReadConsumesZeroRefBases() {
         read.setCigar("60S30I11S");
         Assert.assertEquals(read.getCigar().getReferenceLength(), 0, "read should consume no reference bases");
 
-        Assert.assertFalse(VALID_ALIGNMENT_END.test(read), "Read consuming no reference bases should have failed VALID_ALIGNMENT_END");
+        Assert.assertTrue(VALID_ALIGNMENT_END.test(read), "Read consuming no reference bases should have passes VALID_ALIGNMENT_END");
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibraryUnitTest.java]: typo passes -> passed

**Comment by akiezun** [comment]
@droazen - i enabled those reads to match gatk3

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/bqsr/ApplyBQSRIntegrationTest.java [hunk 4]
```diff
@@ -123,4 +126,19 @@ public void testPRWithConflictingArguments_qqAndRDQ() throws IOException {
                 UserException.CommandLineException.class);
         spec.executeTest("testPRWithConflictingArguments_qqAndSQQ", this);
     }
+
+    @Test
+    public void testOverfiltering() throws IOException {
+        final File zeroRefBasesReadBam = new File(resourceDir, "NA12878.oq.read_consumes_zero_ref_bases.sam");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/bqsr/ApplyBQSRIntegrationTest.java]: Convert to bam

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/bqsr/ApplyBQSRIntegrationTest.java [hunk 5]
```diff
@@ -123,4 +126,19 @@ public void testPRWithConflictingArguments_qqAndRDQ() throws IOException {
...
+        final File outFile = BaseTest.createTempFile("testReadThatConsumesNoReferenceBases", ".sam");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/bqsr/ApplyBQSRIntegrationTest.java]: output to bam

**Comment by droazen** [comment]
:+1: merge after addressing comments

File: src/test/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibraryUnitTest.java (hunk 3 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibraryUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/bqsr/ApplyBQSRIntegrationTest.java (hunk 4 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/bqsr/ApplyBQSRIntegrationTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/bqsr/ApplyBQSRIntegrationTest.java (hunk 5 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/bqsr/ApplyBQSRIntegrationTest.java]: done

**Pull request merged**